Plugins/GitUIPluginInterfaces/ObjectId.cs

```diff
@@ -0,0 +1,178 @@
+
+        public static ObjectId Random()
+        {
+            var bytes = new byte[Sha1ByteCount];
+            _random.NextBytes(bytes);
+            return new ObjectId(bytes);
+        }
+
+        #region Parsing
+
+        public static bool TryParse([CanBeNull] string hex, out ObjectId objectId)
+        {
+            if (hex == null || hex.Length != Sha1CharCount)
+            {
+                objectId = default;
+                return false;
+            }
+
+            return TryParse(hex, 0, out objectId);
+        }
+
+        public static bool TryParse([CanBeNull] string hex, int offset, out ObjectId objectId)
+        {
+            if (hex == null || hex.Length - offset < Sha1CharCount)
+            {
+                objectId = default;
+                return false;
+            }
+
+            var bytes = new byte[Sha1ByteCount];
+
+            var charIndex = 0;
+            for (var i = 0; i < Sha1ByteCount; ++i)
+            {
+                var nibble1 = HexCharToInt(hex[charIndex++]);
+                var nibble2 = HexCharToInt(hex[charIndex++]);
+
+                if (nibble1 < 0 || nibble1 > 15 || nibble2 < 0 || nibble2 > 15)
+                {
+                    objectId = default;
+                    return false;
+                }
+
+                bytes[i] = (byte)((nibble1 << 4) + nibble2);
+            }
+
+            objectId = new ObjectId(bytes);
+            return true;
+
+            int HexCharToInt(char c)
+            {
+                var sub = c <= '9'
+                    ? '0'
+                    : c < 'a'
+                        ? 'A' - 10
+                        : 'a' - 10;
+
+                return c - sub;
+            }
+        }
+
+        public static ObjectId Parse(string hex)
+        {
+            if (hex.Length != Sha1CharCount || !TryParse(hex, 0, out var id))
+                throw new FormatException($"Unable to parse object ID \"{hex}\".");
+            return id;
+        }
+
+        public static ObjectId Parse(string hex, int offset)
+        {
+            if (!TryParse(hex, offset, out var id))
+                throw new FormatException($"Unable to parse object ID \"{hex}\".");
+            return id;
+        }
+
+        #endregion
+
+        public static bool IsValid(string hex)
+        {
+            if (hex.Length != Sha1CharCount)
+                return false;
+
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            // ReSharper disable once ForCanBeConvertedToForeach
+            for (var i = 0; i < hex.Length; i++)
+            {
+                var c = hex[i];
+                if (!char.IsDigit(c) && !(c >= 'a' && c <= 'f') && !(c >= 'A' && c <= 'F'))
+                    return false;
+            }
+
+            return true;
+        }
+
+        [NotNull] private readonly byte[] _bytes;
+
+        private ObjectId([NotNull] byte[] bytes) => _bytes = bytes ?? throw new ArgumentNullException(nameof(bytes));
```

Comment:
I'm tinkering with your proposal (nothing concrete yet, once there is something to share I will push it in) and having few thoughts. 
I'd like to get your opinion.

Do you think we could optimise it by performing `ToString` implementation in the constructor and caching the result in a field, then return the value of the field in `ToString` method?
Does it make sense to calculate the hashcode in constructor as well?

```
        private readonly byte[] _bytes;
        private readonly string _sha1;
        private readonly int _hashCode;

        public ObjectId(IEnumerable<byte> bytes)
        {
            _bytes = bytes?.ToArray() ?? throw new ArgumentNullException(nameof(bytes));

            // If performance here becomes a problem, review https://stackoverflow.com/q/311165/24874
            var hex = new StringBuilder(ObjectIdParser.Sha1CharCount);
            foreach (var b in _bytes)
                hex.AppendFormat("{0:x2}", b);
            _sha1 = hex.ToString();

            // calculate hashcode
            ....
        }
```

Furthermore, having extracted parsing logic into `ObjectIdParser` we can probably cache all known ObjectIds and reuse them, if necessary. This requires further exploration, just a thought.

Reply:
Hehe, brain dump inbound...

Only way to decide on optimisations is to measure. If you want to cache the string then there's no point in storing the `byte[]` too. The string will be roughly four times the size in memory (40 characters, UTF-16), so you'd need context to make tradeoffs between memory/CPU. If you wanted to run such measurements I would look at optimising the existing `ToString` and `GetHashCode` methods first, as they're not the fastest they could be right now. They're already quite snappy, but I wouldn't be surprised if they could be made many times faster still.

I considered caching all object IDs. I've tried/seen attempts at this kind of pooling before and apart from some cases where there are many long-lived instances from a small distinct set of values, there's little benefit and it often tends to come out slower. I would imagine many of the `ObjectId` values would never get promoted from gen0, being collected very soon after construction. Keeping them small and light can help the GC do its job by reducing heap fragmentation and cache pressure. I don't think there's much perf to gain here and there's some potential problems, as well as slight complexity in managing the caches that you avoid if you don't have them. But again, it should be measured. The Linux kernel has ~750k commits. With the 20-byte array, plus 16 bytes overhead for the wrapper class and 12 bytes overhead for the array (minimums on x64), you're looking at ~63MB _before_ you've even begun to store them in something you can look them up from. I just don't think it's going to be a saving. Plus the lookup collection will be in one part of memory/cache, and the wrapper/array in another, so you end up blowing away your CPU caches which tend to be helpful for the work you're actually trying to do.

Any any benefit you'd get on that scale would be offset by representing the byte array as `IEnumerable<byte>` then calling `ToArray` on it again. You want a fast path from parse to object readiness, with minimal allocations. Arrays are about as lean as you'll get without storing the 20 bytes of the SHA-1 in two `ulong`s and an `uint`. The representation is an internal consideration, not a place for abstraction, which is why I made the constructor private.

That's a lot of speculation on my part, especially given I said you one has to measure these things, but it's the kind stuff that would go through my head and lead me towards probably not even doing such comparative measurements until I knew there was a problem they might actually solve.

There are many opportunities to improve the performance and responsivity of GE that would yield orders-of-magnitude better results than such micro-optimisations. For example, have a look how much string concatenation goes on all over the place. These push the GC much harder than needed. And the async stuff I looked at in #4501 is an early (and largely incomplete) attempt at reducing the complete profligacy of creating new threads left and right to deal with blocking code. Running GE in a timeline profiler shows results that look completely different from most other applications because of this. Tonnes of threads doing almost no work.

I'll respond soon to the other comments here that I neglected. Apologies for that.

Reply:
This will decrease the size of GitRevision, but how much part is really the sha?
The Subject, Author etc should use more space already now.
Every byte could count and there may be other reasons for this PR.

For performance I find the unnecessary query of submodules, stash etc to be the most annoying. I have planned to do something about that but have been stuck on in my opinion meaningless test cases for other issues for some time.